clustering/__init__.py

```diff
@@ -10,3 +23,197 @@ def cluster(pd_rules: pd.DataFrame) -> dict:
     data = ~pd_rules.isin([False, 'False', None, 'None', 0])
     clusters = KMeans(n_clusters=5).fit_predict(data)
     return dict(zip(pd_rules['system_id'], clusters))
+
+
+def preprocess_rules_data(DF):
+	
+	# Remove ID and time Columns from training set
+	DF = DF.iloc[:,1:-2] 
+	# Convert any remaining textual data into a nan value (there are very few of these in the dataset)
+	DF = DF.apply(lambda L: pd.to_numeric(L,errors='coerce'))
+	# Fill in any remaning nan values with 0. 
+	DF = DF.fillna(0)
+	return DF
+
+
+def group_clusters(results,num_clusters):
+
+	day = {}
+
+	# initialize our day dictionary with our initial classes
+	for i in range(num_clusters):
+		current_list = []
+		for j in range(len(results)):
+			if i == results[j][1]:
+				current_list.append(results[j][0]) 
+		day[str(i)] = set(current_list)
+
+
+	return day
+
+
+def calculate_stability_score(day_1,day_2,num_clusters,ids_in_both_days):
+
+	stability_score = 0 
+
+	for i in range(num_clusters):
+		smallest_difference = np.inf
+		for j in range(num_clusters):
+			#find best matching cluster:
+			current_difference  = len(day_1[str(i)].difference(day_2[str(j)]))
+			if current_difference < smallest_difference:
+				smallest_difference = current_difference
+		#print(i, " : ", smallest_difference)    
+		stability_score += smallest_difference
+
+	stability_score = ((1-(stability_score/len(ids_in_both_days))) * 100.0 )  
+	#print(str(stability_score) , "%,")
+
+	return stability_score 
+
+
+def run_clustering():
+
+	
+	## Read in our environment variables for storage, tracking, and model ##
+
+	# Storage Parameters
+	ceph_key = os.environ.get("CEPH_KEY")
+	ceph_secret = os.environ.get("CEPH_SECRET")
+	ceph_host = os.environ.get("CEPH_HOST")
+	ceph_bucket = os.environ.get("CEPH_BUCKET")
+
+	# MLFlow Paramters
+	mlflow_experiment_id = os.environ.get('MLFLOW_EXPERIMENT_ID')
+	mlflow_tracking_uri  = os.environ.get("MLFLOW_TRACKING_URI")
+
+	
+	# Model Parameters
+	k_clusters = int(os.environ.get("K_CLUSTERS"))
+	pca_dimensions = int(os.environ.get("PCA_DIMENSIONS"))
```

Comment:
again, how are we getting number of dimensions? There should be code showing variance kept, vs. reducing the dimensions and we should decide based on that?